pkg/plugins/backendplugin/instrumentation/instrumentation.go

```diff
@@ -58,29 +58,127 @@ const (
 	endpointQueryData      = "queryData"
 )
 
+type errorSource string
+
+const (
+	pluginSource   errorSource = "plugin"
+	externalSource errorSource = "external"
+	databaseSource errorSource = "database"
+	noneSource     errorSource = "none"
+)
+
 var logger = plog.New("plugin.instrumentation")
 
 // instrumentPluginRequest instruments success rate and latency of `fn`
 func instrumentPluginRequest(ctx context.Context, cfg Cfg, pluginCtx *backend.PluginContext, endpoint string, fn func() error) error {
-	status := statusOK
-
 	start := time.Now()
 	timeBeforePluginRequest := log.TimeSinceStart(ctx, start)
 
 	err := fn()
-	if err != nil {
-		status = statusError
-		if errors.Is(err, context.Canceled) {
-			status = statusCancelled
+
+	elapsed := time.Since(start)
+	status := getStatus(err)
+
+	logger.Info("Plugin Request Completed")
+
+	updateMetrics(pluginCtx.PluginID, endpoint, string(cfg.Target), elapsed, status, string(noneSource))
+	logDatasourceRequests(ctx, cfg, pluginCtx, endpoint, status, elapsed, timeBeforePluginRequest, err)
+
+	return err
+}
+
+func InstrumentQueryDataRequest(ctx context.Context, cfg Cfg, pluginCtx *backend.PluginContext, requestSize float64, fn func() (*backend.QueryDataResponse, error)) (*backend.QueryDataResponse, error) {
+	start := time.Now()
+	timeBeforePluginRequest := log.TimeSinceStart(ctx, start)
+
+	resp, err := fn()
+
+	elapsed := time.Since(start)
+	logger.Info("Plugin Request Completed")
+
+	status := getStatus(err)
+	errorSource := getErrorSource(status, resp)
+
+	updateMetrics(pluginCtx.PluginID, endpointQueryData, string(cfg.Target), elapsed, status, string(errorSource))
+	pluginRequestSizeHistogram.WithLabelValues("grafana-backend", pluginCtx.PluginID, endpointQueryData, string(cfg.Target)).Observe(requestSize)
+	logDatasourceRequests(ctx, cfg, pluginCtx, endpointQueryData, status, elapsed, timeBeforePluginRequest, err)
+
+	return resp, err
+}
+
+func getErrorSourceForResponse(res backend.DataResponse) errorSource {
+	if res.Error != nil {
+		return pluginSource
+	}
+
+	if res.Status >= 500 {
+		return databaseSource
```

Comment:
Database in this case means the external system the plugin integrates with? This doesn't look bulletproof, it could as well be the plugin failing to return a data query response. Like it depends on how the plugin is implemented and what statuses decides to return. Or am I thinking wrongly about it?

Reply:
You are right that the entire thing depends on how well the database implemented their http api, particularly the http status code.
I seems pretty good for Prometheus, but I know some other plugin would need a custom implementation of `getErrorSource` that is not based on http code.

For the 500, in theory it is supposed to be the database that crashed because of an issue from their side.
In practice there is a lot of services that return 500 instead of 400 if you don't send them valid data.

I don't think that there is a bulletproof solution, this one included. The idea is to first try with something that in theory, most of the time, should report the right thing, and then when we start receiving the data, and see manually if that is useful or not etc, then we iterate and improve the logic